### PR TITLE
Add ability to copy internal link from share dialog

### PR DESCRIPTION
**Pull request by claucambra**
This addresses #815 and helps replicate the functionality available on the web interface

image

### CODE REVIEW

File: src/gui/internallinkwidget.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    : QWidget(parent)
+    , _ui(new Ui::InternalLinkWidget)
+    , _account(account)
+    , _localPath(localPath)
+{
+    _ui->setupUi(this);
+
+    //Is this a file or folder?
+    const QFileInfo fi(localPath);
+    _isFile = fi.isFile();
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:41]: _isFile is not being used anywhere.

File: src/gui/internallinkwidget.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+void InternalLinkWidget::slotCopyInternalLink(const bool clicked) const
+{
+    Q_UNUSED(clicked)
+
+    QApplication::clipboard()->setText(_internalUrl);
+}
+
+void InternalLinkWidget::slotStyleChanged()
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:86]: this slot is not being used.

File: src/gui/internallinkwidget.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#include "theme.h"
+
+#include "QProgressIndicator.h"
+#include <QClipboard>
+#include <QToolButton>
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcInternalLink, "nextcloud.gui.internaklink", QtInfoMsg)
+
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:28]: nextcloud.gui.internaklink => nextcloud.gui.internallink

File: src/gui/internallinkwidget.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    , _localPath(localPath)
+{
+    _ui->setupUi(this);
+
+    //Is this a file or folder?
+    const QFileInfo fi(localPath);
+    _isFile = fi.isFile();
+
+    const auto folder = FolderMan::instance()->folderForPath(localPath);
+    const auto folderRelativePath = localPath.mid(folder->cleanPath().length() + 1);
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:44]: localPath => _localPath

File: src/gui/internallinkwidget.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcInternalLink, "nextcloud.gui.internaklink", QtInfoMsg)
+
+InternalLinkWidget::InternalLinkWidget(AccountPtr account,
+    const QString &localPath,
+    QWidget *parent)
+    : QWidget(parent)
+    , _ui(new Ui::InternalLinkWidget)
+    , _account(account)
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:34]: _account is not being used.

File: src/gui/internallinkwidget.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    const QFileInfo fi(localPath);
+    _isFile = fi.isFile();
+
+    const auto folder = FolderMan::instance()->folderForPath(localPath);
+    const auto folderRelativePath = localPath.mid(folder->cleanPath().length() + 1);
+    const auto serverRelativePath = QDir(folder->remotePath()).filePath(folderRelativePath);
+
+    SyncJournalFileRecord record;
+    if (folder)
+        folder->journalDb()->getFileRecord(folderRelativePath, &record);
```
[the remark this answers is not in the document]
Comment (camilasan) [reply src/gui/internallinkwidget.cpp:49]: and I don't think you need to call getFileRecord here.

File: src/gui/internallinkwidget.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+{
+    Q_OBJECT
+
+public:
+    explicit InternalLinkWidget(AccountPtr account,
+        const QString &localPath,
+        QWidget *parent = nullptr);
+    ~InternalLinkWidget() = default;
+
+    void toggleButton(bool show);
```
Comment (camilasan) [inline src/gui/internallinkwidget.h:41]: toggleButton is not implemented.

**Review by camilasan — CHANGES_REQUESTED** [review]

**Comment by camilasan** [comment]
Locally I can not compile it. I get the same error as SonarCloud.

File: src/gui/internallinkwidget.h (hunk 7 above)
Comment (claucambra) [reply src/gui/internallinkwidget.h]: Removed

File: src/gui/internallinkwidget.cpp (hunk 3 above)
Comment (claucambra) [reply src/gui/internallinkwidget.cpp:28]: Oops! Fixed the typo

File: src/gui/internallinkwidget.cpp (hunk 2 above)
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: Removed this slot

File: src/gui/internallinkwidget.cpp (hunk 1 above)
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: True, removed

File: src/gui/internallinkwidget.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: I added the braces and you;re right, don't need to call getFileRecord, removed

**Comment by claucambra** [comment]
> Locally I can not compile it. I get the same error as SonarCloud.

Made a bad use of QScopedPointer, fixed now :)

File: src/gui/internallinkwidget.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+#include "folderman.h"
+#include "theme.h"
+
+#include "QProgressIndicator.h"
+#include <QClipboard>
+#include <QToolButton>
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcInternalLink, "nextcloud.gui.internalink", QtInfoMsg)
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:27]: internalink => internallink (like in the class name)

File: src/gui/internallinkwidget.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    connect(_ui->copyInternalLinkButton, &QPushButton::clicked, this, &InternalLinkWidget::slotCopyInternalLink);
+}
+
+InternalLinkWidget::~InternalLinkWidget()
+{
+    delete _ui;
+}
+
+void InternalLinkWidget::setupUiOptions()
```
Comment (camilasan) [inline src/gui/internallinkwidget.cpp:64]: this function is not being used.

**Review by camilasan — CHANGES_REQUESTED** [review]

File: src/gui/internallinkwidget.cpp (hunk 9 above)
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: Thanks, removed it

File: src/gui/internallinkwidget.cpp (hunk 8 above)
Comment (claucambra) [reply src/gui/internallinkwidget.cpp:25]: Nice catch, fixed

File: src/gui/internallinkwidget.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    void setupUiOptions();
+
+private slots:
+    void slotLinkFetched(const QString &url);
+    void slotCopyInternalLink(const bool clicked) const;
+
+private:
+    void customizeStyle();
+
+    Ui::InternalLinkWidget *_ui;
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.h]: Yes, this is cleaner -- changed, thanks

File: src/gui/internallinkwidget.cpp [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    _ui->copyInternalLinkButton->setEnabled(false);
+    _ui->internalLinkProgressIndicator->setVisible(true);
+    _ui->internalLinkProgressIndicator->startAnimation();
+
+    connect(_ui->copyInternalLinkButton, &QPushButton::clicked, this, &InternalLinkWidget::slotCopyInternalLink);
+}
+
+InternalLinkWidget::~InternalLinkWidget()
+{
+    delete _ui;
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: This is now removed

File: src/gui/internallinkwidget.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    , _ui(new Ui::InternalLinkWidget)
+    , _localPath(localPath)
+{
+    _ui->setupUi(this);
+
+    const auto folder = FolderMan::instance()->folderForPath(_localPath);
+    const auto folderRelativePath = _localPath.mid(folder->cleanPath().length() + 1);
+    const auto serverRelativePath = QDir(folder->remotePath()).filePath(folderRelativePath);
+
+    SyncJournalFileRecord record;
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp:39]: Looks like nothing, as `numericFileId` converts the QByteArray `fileId` to integer iteratively -- probably a better idea to just provide an empty object to `fetchPrivateLinkUrl`, which already handles receiving an empty `record.numericFileId()`

[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: Did this

File: src/gui/internallinkwidget.cpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+{
+    _internalUrl = url;
+    _ui->copyInternalLinkButton->setEnabled(true);
+    _ui->internalLinkProgressIndicator->setVisible(false);
+    _ui->internalLinkProgressIndicator->stopAnimation();
+}
+
+void InternalLinkWidget::slotCopyInternalLink(const bool clicked) const
+{
+    Q_UNUSED(clicked);
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: You're right, removed the arg

**Comment by mgallien** [comment]
@claucambra can you have a look at the conflict and rebase on your side ?

File: src/gui/sharedialog.cpp [hunk 14]
```diff
@@ -135,13 +136,17 @@ ShareDialog::ShareDialog(QPointer<AccountState> accountState,
     connect(job, &PropfindJob::result, this, &ShareDialog::slotPropfindReceived);
     connect(job, &PropfindJob::finishedWithError, this, &ShareDialog::slotPropfindError);
     job->start();
-
+
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/sharedialog.cpp]: Nice catch, fixed

File: src/gui/internallinkwidget.cpp [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+#include <QClipboard>
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcInternalLink, "nextcloud.gui.internalink", QtInfoMsg)
+
+InternalLinkWidget::InternalLinkWidget(const QString &localPath,
+    QWidget *parent)
+    : QWidget(parent)
+    , _ui(new Ui::InternalLinkWidget)
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/internallinkwidget.cpp]: Thanks, used make_unique and moved it to the class declaration to fix code smell

File: src/gui/sharedialog.cpp [hunk 16]
```diff
@@ -135,13 +136,17 @@ ShareDialog::ShareDialog(QPointer<AccountState> accountState,
...
-
+      
     initShareManager();
     
     _scrollAreaViewPort = new QWidget(_ui->scrollArea);
     _scrollAreaLayout = new QVBoxLayout(_scrollAreaViewPort);
     _scrollAreaLayout->setContentsMargins(0, 0, 0, 0);
     _ui->scrollArea->setWidget(_scrollAreaViewPort);
+
+
```
[the remark this answers is not in the document]
Comment (claucambra) [reply src/gui/sharedialog.cpp]: You;re right, removed second space

**Review by jancborchardt — CHANGES_REQUESTED** [review]
2 things design-wise:
- The icon we use in server is different from the normal link one to separate them: [icon-external](https://docs.nextcloud.com/server/22/developer_manual/design/icons.html#id35)
- The text would look better if the subline is indented the same as the main text, and they are vertically aligned as a unit with the icons (as in the server)

See server for reference:
image

**Comment by claucambra** [comment]
> 2 things design-wise:
> 
>     * The icon we use in server is different from the normal link one to separate them: [icon-external](https://docs.nextcloud.com/server/22/developer_manual/design/icons.html#id35)
> 
>     * The text would look better if the subline is indented the same as the main text, and they are vertically aligned as a unit with the icons (as in the server)
> 
> 
> See server for reference: image

I followed your suggestions and tried to make the widget look as close to the server's as possible:

image

**Review by jancborchardt — APPROVED** [review]
Much better @claucambra, thank you! :)

File: src/gui/internallinkwidget.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+ * @ingroup gui
+ */
+class InternalLinkWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    explicit InternalLinkWidget(const QString &localPath,
+        QWidget *parent = nullptr);
+    ~InternalLinkWidget() = default;
```
Comment (mgallien) [inline src/gui/internallinkwidget.h:37]: you are missing an `override` because you override the destructor of `QWidget` class

Comment (mgallien) [reply src/gui/internallinkwidget.h:37]: funny thing is that adding this comment made me look at the drone CI report that is indeed broken and requests exactly the same [Drone CI check](https://drone.nextcloud.com/nextcloud/desktop/5231/2/5)

**Review by mgallien — CHANGES_REQUESTED** [review]

File: src/gui/internallinkwidget.h (hunk 17 above)
Comment (claucambra) [reply src/gui/internallinkwidget.h]: Okay, added the `override` :)

**Comment by mgallien** [comment]
@claucambra the drone check is still failing because there was also another failed static check on code
can you have a look ?

**Comment by claucambra** [comment]
> @claucambra the drone check is still failing because there was also another failed static check on code can you have a look ?

Right, thanks, fixed it

**Review by mgallien — APPROVED** [review]

**Review by camilasan — APPROVED** [review]

**Pull request merged**